### DISCUSSION TITLE
Caching memoized results in React Query

### DISCUSSION THREAD

**Discussion by toppsdown** [TanStack/query#7767] [Q&A] [answered]
I've been toying with an idea but not sure if it's a good one. Unfortunately, I haven't been able to find anything on google about it so hope ya'll can tell me if I'm being stupid.

I'm using useQuery to grab a set of data from the backend, but then I need to manipulate that data on the frontend to transform it into a few different views. I'd like to avoid re-calculating unless the returned data has changed.

I first reached for useMemo, but realized that would recalculate the data on each unmount/remount which is unnecessary.

So I've been thinking about storing the transformed data in react query so any component can access the cached data. The transformed data would have an infinite staleTime and would only get recalculated if useQuery returns new data from the backend.

My guess is that a state management library would be a better tool for the job, but we don't currently have one in the codebase. React query has been sufficient for us so far.

Thoughts?
Thanks in advance for any advice!

**Comment by maddyahmad** [comment]
Redux is always a better option. If you don't wanna use that, maybe use select option in useQuery, convert the response into the view you want, and return that object and the data you will get from the api will return that object only.

const {
data: comments,
} = useQuery({
    queryKey: ['comments'],
    queryFn: async () => getInstance(GET_COMMENTS_URL),
    staleTime: 1000 * 60 * 5,
    select: (data) => {
      dispatch(saveCommentsData(data.comments))
      return data.comments
    },
  });

**Reply by toppsdown** [reply]
Can you elaborate on why Redux is a better option than react query for storing processed data?

I need to keep the raw data and also create 3 transformed versions for different app views. Seems messy to me to store all that under the same query key.

**Reply by TkDodo** [reply]
sorry to say but dispatching data from react-query to redux is an absolute anti pattern. Please see:

https://tkdodo.eu/blog/thinking-in-react-query

**Reply by toppsdown** [reply]
> sorry to say but dispatching data from react-query to redux is an absolute anti pattern. Please see:
> 
> https://tkdodo.eu/blog/thinking-in-react-query

Great, that's helpful. Thank you!

**Comment by TkDodo** [comment, marked as answer]
>  but then I need to manipulate that data on the frontend to transform it into a few different views. I'd like to avoid re-calculating unless the returned data has changed.

you are likely looking for the `select` option:

- https://tanstack.com/query/latest/docs/framework/react/guides/render-optimizations#select
- https://tkdodo.eu/blog/react-query-data-transformations#3-using-the-select-option

**Reply by toppsdown** [reply]
Ah, I had read the select option before but didn't realize you could have multiple selects for the same underlying data attached to a single query key, and that it only runs on data update.

That's perfect, thank you! And thanks for creating such a great library. It's been a joy to work with.
